### PR TITLE
Implementation of Pallas

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

File: orchard_pallas.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            return a
+        elif a.is_identity:
+            return self
+        else:
+            (x1, y1) = (self.x, self.y)
+            (x2, y2) = (a.x, a.y)
+
+            if x1 != x2:
+                # <https://core.ac.uk/download/pdf/10898289.pdf> section 4.1
+                λ = (y1 - y2) / (x1 - x2)
```
Comment (dconnolly) [inline orchard_pallas.py:167]: What version of Python is required to parse these lambda chars?

Comment (daira) [reply orchard_pallas.py:167]: Actually this works on Python 2.7 with the `# -*- coding: utf8 -*-` line. But this file is explicitly declared as Python 3 anyway.

**Comment by str4d** [comment]
This PR now just covers the Pallas implementation. Test vectors are in subsequent PRs like #17.

File: orchard_pallas.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+assert (p - 1) // 2 == pm1d2
+
+S = 32
+T = 0x40000000000000000000000000000000224698fc094cf91b992d30ed
+assert (p - 1) == (1 << S) * T
+
+tm1d2 = 0x2000000000000000000000000000000011234c7e04a67c8dcc969876
+assert (T - 1) // 2 == tm1d2
+
+ROOT_OF_UNITY = 0x2bce74deac30ebda362120830561f81aea322bf2b7bb7584bdad6fabd87ea32f
```
Comment (daira) [inline orchard_pallas.py:21]:
```suggestion
// 5^T (mod p)
ROOT_OF_UNITY = 0x2bce74deac30ebda362120830561f81aea322bf2b7bb7584bdad6fabd87ea32f
```

**Review by daira — APPROVED** [review]
ACK

File: orchard_pallas.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+        if y.s % 2 != y_sign:
+            y = Fp.ZERO - y
+
+        return Point(x, y)
+
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+        self.is_identity = False
```
Comment (ebfull) [inline orchard_pallas.py:144]: Would be cool to have an is_on_curve assertion here; `y^2 == x^3 + b`.

File: orchard_pallas.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            return 'Point(%s, %s)' % (self.x, self.y)
+
+
+Point.ZERO = Point.identity()
+Point.GENERATOR = Point(Fp.MINUS_ONE, Fp(2))
+
+assert Point.ZERO + Point.ZERO == Point.ZERO
+assert Point.GENERATOR - Point.GENERATOR == Point.ZERO
+assert Point.GENERATOR + Point.GENERATOR + Point.GENERATOR == Point.GENERATOR * Scalar(3)
+assert Point.GENERATOR + Point.GENERATOR - Point.GENERATOR == Point.GENERATOR
```
Comment (ebfull) [inline orchard_pallas.py:229]: A good test is `assert (-Point.GENERATOR) == Point.GENERATOR * -Scalar(-1)`.

**Review by ebfull — APPROVED** [review]
ACK with non-blocking comments/nits.

**Comment by str4d** [comment]
Will address @ebfull's nits in a follow-up PR, since I don't trust GitHub to switch the base of #17 correctly if I push additional commits to this PR.

**Pull request merged**